scripts/package

```diff
@@ -5,21 +5,52 @@ source $(dirname $0)/version
 
 cd $(dirname $0)/..
 
-PROJECT=`basename "$PWD"`
+PROJECT=$(basename "$PWD")
 
-if [ ! -x ./bin/longhorn ]; then
-    ./scripts/build
+command -v buildx >/dev/null && BUILD_CMD=(buildx) || BUILD_CMD=(docker buildx)
+
+# read configurable parameters
+REPO=${REPO:-longhornio}
+IMAGE_NAME=${IMAGE_NAME:-$PROJECT}
+TAG=${TAG:-''}
+OUTPUT_ARGS=${OUTPUT_ARGS:-'--load'}
+IS_SECURE=${IS_SECURE:-'false'}
+MACHINE=${MACHINE:-''}
+TARGET_PLATFORMS=${TARGET_PLATFORMS:-''}
+IID_FILE=${IID_FILE:-''}
+IID_FILE_FLAG=${IID_FILE_FLAG:-''}
+
+if [[ -z "$TAG" ]]; then
+    if API_VERSION=$(./bin/backing-image-manager version --client-only | jq ".clientVersion.backingImageManagerAPIVersion"); then
+      TAG="v${API_VERSION}_$(date -u +%Y%m%d)"
+    else
+      TAG="${VERSION}"
+    fi
 fi
 
-cp -r bin package/
+IMAGE="${REPO}/${IMAGE_NAME}:${TAG}"
 
-APIVERSION=`./bin/backing-image-manager version --client-only|jq ".clientVersion.backingImageManagerAPIVersion"`
-TAG="v${APIVERSION}_`date -u +%Y%m%d`"
-REPO=${REPO:-longhornio}
-IMAGE=${REPO}/${PROJECT}:${TAG}
+BUILDER_ARGS=()
+[[ ${MACHINE} ]] && BUILDER_ARGS+=('--builder' "${MACHINE}")
+
+IFS=' ' read -r -a IID_FILE_ARGS <<<"${IID_FILE_FLAG}"
+[[ -n "${IID_FILE}" && ${#IID_FILE_ARGS} == 0 ]] && IID_FILE_ARGS=('--iidfile' "${IID_FILE}")
+
+IFS=' ' read -r -a BUILDX_ARGS <<<"${OUTPUT_ARGS}"
+[[ ${IS_SECURE} == 'true' ]] && BUILDX_ARGS+=('--sbom=true' '--attest' 'type=provenance,mode=max')
+[[ ${TARGET_PLATFORMS} ]] && BUILDX_ARGS+=('--platform' "${TARGET_PLATFORMS}")
 
-buildx build --load --no-cache -t ${IMAGE} -f package/Dockerfile .
+IMAGE_BUILD_CMD=(
+    "${BUILD_CMD[@]}" build --no-cache \
+    "${BUILDER_ARGS[@]}" \
+    "${IID_FILE_ARGS[@]}" \
+    "${BUILDX_ARGS[@]}" \
+    -t "${IMAGE}" -f package/Dockerfile .
+)
+echo "${IMAGE_BUILD_CMD[@]}"
+"${IMAGE_BUILD_CMD[@]}"
 
-echo Built ${IMAGE}
+echo "Built ${IMAGE}"
 
-echo ${IMAGE} > ./bin/latest_image
+mkdir ./bin || true
+echo "${IMAGE}" > ./bin/latest_image
```

Comment:
What's the purpose to dump another image here? For debugging? If yes, please make a flag to enable it and disable it by default, as the image will be pushed to the registry instead.

Reply:
We record the output image name into this text file for further automation.